### PR TITLE
Add tests for dynamic cache mode switching

### PR DESCRIPTION
**Pull request by rafalste**
Signed-off-by: Rafal Stefanowski <rafal.stefanowski@intel.com>

### CODE REVIEW

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+from test_utils.os_utils import Udev, sync
+from test_tools.fio.fio import Fio
+from test_tools.fio.fio_param import ReadWrite, IoEngine
+
+
+io_size = Size(10000, Unit.Blocks4096)
+
+
+@pytest.mark.parametrize(
+    "cache_mode",
```
Comment (micrakow) [inline test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: general - how about write-only cache mode?

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+    with TestRun.step("Stop 'fio'"):
+        TestRun.executor.kill_process(fio_pid)
+
+    with TestRun.step("Stop all caches"):
+        casadm.stop_all_caches()
+
+
+def storage_prepare():
+    cache_dev = TestRun.disks["cache"]
+    cache_dev.create_partitions([Size(2, Unit.GibiByte)])
```
Comment (micrakow) [inline test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Is it intended that cache is bigger than core?

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+def run_io_and_verify(cache, core, io_mode, io_size):
+    fio_prepare(core, io_mode, io_size).run()
+    sync()
+    cache_mode = cache.get_cache_mode()
+    cache_stats = cache.get_statistics()
+    core_stats = core.get_statistics()
+
+    if cache_mode == CacheMode.WB:
+        if core_stats.block_stats.core.writes < core_stats.block_stats.exp_obj.writes:
+            TestRun.LOGGER.info("Write-Back cache mode works properly")
```
Comment (micrakow) [inline test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: General - do we need this kind of logs in tests? Don't we expect everything to work fine and log only if something is wrong? What do You think?

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+            if check_cache_config.cache_mode != cache.get_cache_mode():
+                TestRun.fail("Cache mode do not match after reload!")
+            if check_cache_config.cleaning_policy != cache.get_cleaning_policy():
+                TestRun.fail("Cleaning policy do not match after reload!")
+            if check_cache_config.eviction_policy != cache.get_eviction_policy():
+                TestRun.fail("Eviction policy do not match after reload!")
+            if check_cache_config.cache_line_size != cache.get_cache_line_size():
+                TestRun.fail("Cache line size do not match after reload!")
+            if check_cache_config.metadata_mode != cache.get_metadata_mode():
+                TestRun.fail("Metadata mode do not match after reload!")
```
Comment (micrakow) [inline test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: How about reporting all those mismatches at once? Right now if many of properties do not match we do not log all of them and just fail on the first difference instead.

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+@pytest.mark.require_disk("cache", DiskTypeSet([DiskType.optane, DiskType.nand]))
+@pytest.mark.require_disk("core", DiskTypeLowerThan("cache"))
+def test_cache_mode_switching_during_io(cache_mode_1, cache_mode_2, flush, io_mode):
+    """
+        title: Test for dynamic cache mode switching during IO.
+        description: |
+          Validate the ability of the CAS to switch cache modes
+          during working IO on CAS device.
+        pass_criteria:
+          - Cache mode is switched without errors.
```
Comment (micrakow) [inline test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: How about checking data integrity as well?

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+
+    with TestRun.step("Add core to the cache"):
+        core = cache.add_core(core_dev)
+
+    with TestRun.step("Run 'fio'"):
+        fio = (
+            Fio()
+            .create_command()
+            .io_engine(IoEngine.libaio)
+            .block_size(Size(4, Unit.KibiByte))
```
Comment (micrakow) [inline test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Is it some special value?

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+    with TestRun.step(
+        f"Change cache mode to {cache_mode_2} with flush cache set to: {flush}"
+    ):
+        cache_mode_switch_output = cache.set_cache_mode(cache_mode_2, flush)
+
+    with TestRun.step(f"Check if cache mode has switched properly during IO"):
+        cache_mode_after_switch = cache.get_cache_mode()
+
+        if (
+            cache_mode_switch_output.exit_code == 0
```
Comment (micrakow) [inline test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: First check could be moved to the [new] step before this

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+        cache_dev, core_dev = storage_prepare()
+
+    with TestRun.step(f"Start cache in {cache_mode_1} mode"):
+        cache = casadm.start_cache(cache_dev, cache_mode_1, force=True)
+
+    with TestRun.step("Add core to the cache"):
+        core = cache.add_core(core_dev)
+
+    with TestRun.step("Run 'fio'"):
+        fio = (
```
Comment (micrakow) [inline test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: how about using fio_prepare helper instead?

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+                    "'cache writes' and 'core reads' should equal data reads."
+                )
+
+
+def fio_prepare(core, io_mode, io_size):
+    fio = (
+        Fio()
+        .create_command()
+        .io_engine(IoEngine.libaio)
+        .block_size(Size(4, Unit.KibiByte))
```
Comment (micrakow) [inline test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Is it some special value?

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+    "cache_mode_1,cache_mode_2,flush",
+    [
+        (CacheMode.WT, CacheMode.WB, False),
+        (CacheMode.WB, CacheMode.PT, False),
+        (CacheMode.WB, CacheMode.PT, True),
+        (CacheMode.WB, CacheMode.WT, False),
+        (CacheMode.WB, CacheMode.WT, True),
+        (CacheMode.PT, CacheMode.WT, False),
+        (CacheMode.WT, CacheMode.WA, False),
+    ],
```
Comment (Ostrokrzew) [inline test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: As above - You forgot about write-only mode.

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+                and cache_stats.block_stats.core.reads.value == io_size.value
+            ):
+                TestRun.LOGGER.info(
+                    "Write-Around cache mode works properly for data reads"
+                )
+            else:
+                TestRun.fail(
+                    "Write-Around cache mode is not working properly for data reads! "
+                    "'cache writes' and 'core reads' should equal data reads."
+                )
```
Comment (Ostrokrzew) [inline test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Consider using 'switch-case' statement, may be more readable in this case.

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py (hunk 1 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: done

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py (hunk 2 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Changed to bigger core than cache.

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py (hunk 3 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Removed

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py (hunk 4 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Done

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py (hunk 5 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Done

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py (hunk 6 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Removed

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py (hunk 7 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Done

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py (hunk 8 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Done

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py (hunk 9 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Removed

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py (hunk 10 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Done

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 12]
```diff
@@ -0,0 +1,370 @@
...
+    ],
+)
+@pytest.mark.parametrize("io_mode", [ReadWrite.randwrite, ReadWrite.randrw])
+@pytest.mark.require_disk("cache", DiskTypeSet([DiskType.optane, DiskType.nand]))
+@pytest.mark.require_disk("core", DiskTypeLowerThan("cache"))
+def test_cache_mode_switching_during_io(cache_mode_1, cache_mode_2, flush, io_mode):
+    """
+        title: Test for dynamic cache mode switching during IO.
+        description: |
+          Validate the ability of the CAS to switch cache modes
```
Comment (Deixx) [inline test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: CAS w/o 'the'

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 13]
```diff
@@ -0,0 +1,370 @@
...
+            fio_prepare(core, io_mode)
+            .verify(VerifyMethod.sha1)
+            .run_time(timedelta(minutes=4))
+            .time_based()
+        )
+        fio_pid = fio.run_in_background()
+        time.sleep(5)
+
+    with TestRun.step(
+        f"Change cache mode to {cache_mode_2} with flush cache set to: {flush}"
```
Comment (Deixx) [inline test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: with '--flush-cache' set to

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 14]
```diff
@@ -0,0 +1,370 @@
...
+        ):
+            TestRun.fail(
+                "Write-Back cache mode is not working properly! "
+                "There should be some writes to CAS device and none to the core."
+            )
+
+    if cache_mode == CacheMode.PT:
+        if (
+            cache_stats.request_stats.read.total != 0
+            and cache_stats.request_stats.write.total != 0
```
Comment (Deixx) [inline test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: or

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 15]
```diff
@@ -0,0 +1,370 @@
...
+        ):
+            TestRun.fail(
+                "Pass-Through cache mode is not working properly! "
+                "There should be no reads or writes to cache."
```
Comment (Deixx) [inline test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: from/to

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 16]
```diff
@@ -0,0 +1,370 @@
...
+            TestRun.fail(
+                "Write-Through cache mode is not working properly! "
+                "'cache writes' and 'core writes' counts should be the same."
+            )
+
+    if cache_mode == CacheMode.WA:
+        if io_mode == ReadWrite.randread:
+            if (
+                cache_stats.block_stats.cache.writes != io_size
+                and cache_stats.block_stats.core.reads != io_size
```
Comment (Deixx) [inline test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: or

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 17]
```diff
@@ -0,0 +1,370 @@
...
+            ):
+                TestRun.fail(
+                    "Write-Around cache mode is not working properly for data reads! "
+                    "'cache writes' and 'core reads' should equal total data reads."
+                )
+        if io_mode == ReadWrite.randwrite:
+            if cache_stats.block_stats.cache.writes != io_size:
```
Comment (Deixx) [inline test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: != 0

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 18]
```diff
@@ -0,0 +1,370 @@
...
+    fio_prepare(core, io_mode, io_size_after_reload).run()
+    sync()
+    cache_stats = cache.get_statistics()
+    core_stats = core.get_statistics()
+    io_new_data = io_size_after_reload - io_size
+
+    if cache_mode == CacheMode.WA:
+        if (
+            cache_stats.block_stats.cache.writes != io_new_data
+            and cache_stats.block_stats.core.reads != io_new_data
```
Comment (Deixx) [inline test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: or

**Comment by katlapinka** [comment]
@CAS-Linux-Jenkins retest this please

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 19]
```diff
@@ -0,0 +1,370 @@
...
+
+    cache.reset_counters()
+    io_mode = ReadWrite.randwrite
+    fio_prepare(core, io_mode, io_size_after_reload).run()
+    sync()
+    cache_stats = cache.get_statistics()
+    core_stats = core.get_statistics()
+
+    if cache_mode == CacheMode.WA:
+        if cache_stats.block_stats.cache.writes != io_size_after_reload:
```
Comment (Deixx) [inline test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: != 0

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 20]
```diff
@@ -0,0 +1,370 @@
...
+
+def storage_prepare():
+    cache_dev = TestRun.disks["cache"]
+    cache_dev.create_partitions([Size(1, Unit.GibiByte)])
+    core_dev = TestRun.disks["core"]
+    core_dev.create_partitions([Size(2, Unit.GibiByte)])
+    return cache_dev.partitions[0], core_dev.partitions[0]
+
+
+def check_cache_mode_operation(cache, core, io_size, cache_mode):
```
[the remark this answers is not in the document]
Comment (rafalste) [reply test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Done

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py [hunk 21]
```diff
@@ -0,0 +1,370 @@
...
+    if io_size_after_reload <= io_size:
+        TestRun.fail(
+            "io_size_after_reload value is not greater then global io_size value!"
+        )
+
+    io_mode = ReadWrite.randread
+    fio_prepare(core, io_mode, io_size_after_reload).run()
+    sync()
+    cache_stats = cache.get_statistics()
+    core_stats = core.get_statistics()
```
[the remark this answers is not in the document]
Comment (rafalste) [reply test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Done

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py (hunk 12 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Done

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py (hunk 13 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Done

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py (hunk 14 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Done

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py (hunk 15 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Done

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py (hunk 16 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Done

File: test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py (hunk 18 above)
Comment (rafalste) [reply test/functional/tests/cache_ops/test_dynamic_cache_mode_switching.py]: Done

**Pull request merged**